cel/README.md

```diff
@@ -37,231 +48,357 @@ kubectl apply --filename https://storage.cloud.google.com/tekton-releases-nightl
+        apiVersion: custom.tekton.dev/v1alpha1
         kind: CEL
-      params:
-        - name: red
-          value: "{'blue': '0x000080', 'red': '0xFF0000'}['red']"
+        name: get-red
 ```
-### Configuring a `CEL` `Custom Task` in a `PipelineRun`
+### Configuring a `CEL` in a `PipelineRun`
 
-The `CEL` `Custom Task` can be specified within a `PipelineRun`, as such:
+A `CEL` can be specified within a `PipelineRun`, as such:
 
 ```yaml
+apiVersion: custom.tekton.dev/v1alpha1
+kind: CEL
+metadata:
+  name: get-blue
+spec:
+  expressions:
+    - name: expression
+      value: "{'blue': '0x000080', 'red': '0xFF0000'}['blue']"
+---
 apiVersion: tekton.dev/v1beta1
 kind: PipelineRun
 metadata:
-  generateName: pipelinerun-
+  generateName: pipelinerun-get-blue-
 spec:
   pipelineSpec:
     tasks:
       - name: get-blue
         taskRef:
-          apiVersion: cel.tekton.dev/v1alpha1
+          apiVersion: custom.tekton.dev/v1alpha1
           kind: CEL
-        params:
-          - name: blue
-            value: "{'blue': '0x000080', 'red': '0xFF0000'}['blue']"
+          name: get-blue 
 ```
 
 ### Specifying CEL expressions
 
-The CEL expressions to be evaluated by the `Run` are specified using parameters. The parameters can be specified
-in the `Run` directly or be passed through from a `Pipeline` or `PipelineRun`, as such:
+The CEL expressions to be evaluated by the `CEL` are specified in `expressions` field, which are made up of `name` and 
+`value` pairs which are `Strings`.
+
+```yaml
+apiVersion: custom.tekton.dev/v1alpha1
+kind: CEL
+metadata:
+  name: get-type
+spec:
+  expressions:
+    - name: type
+      value: "type(1)"
+---
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: pipeline-run-get-type-
+spec:
+  serviceAccountName: 'default'
+  pipelineSpec:
+    tasks:
+      - name: get-type
+        taskRef:
+          apiVersion: custom.tekton.dev/v1alpha1
+          kind: CEL
+          name: get-type
+      - name: echo-get-type
+        when:
+          - input: "$(tasks.get-type.results.type)"
+            operator: in
+            values: ["int"]
+        taskSpec:
+          steps:
+            - name: echo
+              image: ubuntu
+              script: echo ISINT!
+```
+
+For more information about `Parameters`, read [specifying `Parameters`][specifying-parameters].
+
+### Specifying CEL environment variables
+
+For each execution of a `CEL`, we create an evaluation environment. As described in [CEL Evaluation Documentation][cel-evaluation], 
+context variables can be bound to the environment. 
+
+The CEL variables to be declared in the environment before evaluation by `CEL` are specified in `variables` field, which
+are made up of `name` and `value` pairs which are `Strings`.
 
 ```yaml
+apiVersion: custom.tekton.dev/v1alpha1
+kind: CEL
+metadata:
+  name: get-sev
+spec:
+  variables:
+    - name: priority
+      value: "high"
+    - name: alert_enable
+      value: "true"
```

Comment:
does anything special happen with the values of these variables? i assume they are themselves evaluated? e.g. looking at https://github.com/tektoncd/experimental/issues/716 it looks like the values of these variables could themselves be expressions, is that right?

if so i think it might be worth explaining that a bit in the docs - but also id like to understand more about what the line is between variables and expressions  - e.g. could i use the body of the expression below as the value of one of these variables? `"priority in ['high', 'normal'] ? 'sev-1' : 'sev-2'"` (and do we then hit the same problem of referring to other variables within these variables that was in the original issue?)

Reply:
no, the values of the variables are not evaluated -- they just get defined in the evaluation environment as context variables

in https://github.com/tektoncd/experimental/issues/716, the author had requested that each expression is evaluated then set as an environment variable 

```yaml
params:
- name: name
  value: "'hello'"
- name: types
  value: "type(name)"
```

but I had concerns about that approach discussed in https://github.com/tektoncd/experimental/pull/717#pullrequestreview-606410727 and https://github.com/tektoncd/experimental/pull/717#issuecomment-817918600

here, we are explicit in specifying the variables and expressions separately -- so the example above would be written as such instead:

```yaml
variables:
- name: name
  value: "hello"
expressions:
- name: types
  value: "type(name)"
```

if needed, a user can reuse the result of a previous CELRun as a variable in a subsequent CELRun

Reply:
ah okay i think i finally understand - I was thinking of `variables` behaving the same way as `params` do today, but actually `variables` will be evaluated as CEL environment vars vs. `params`  which were just string replacements 👍